File: packages/distance/src/token/sorensen.ts
import {
  charFrequencyMap,
  frequencyMap,
  intersectCount,
  totalCount,
  ngrams,
  ngramFrequencyMap,
  intersectCountInt,
  totalCountInt,
  CHAR_FREQ_SIZE,
  buildCharFreqArray,
} from "../utils";

// Reusable buffers to avoid per-call allocation
const _freqA = new Int32Array(CHAR_FREQ_SIZE);
const _freqB = new Int32Array(CHAR_FREQ_SIZE);

/**
 * Sørensen-Dice coefficient between two strings based on character-level multiset.
 *
 * DSC(A, B) = 2 * |A ∩ B| / (|A| + |B|)
 *
 * Uses Counter (frequency map) for multiset semantics,
 * matching the textdistance crate behavior.
 *
 * Time: O(m + n)
 *
 * @param a - First string
 * @param b - Second string
 * @returns Sørensen-Dice coefficient in [0, 1]
 */
export function sorensen(a: string, b: string): number {
  // ASCII fast path: Int32Array instead of Map
  _freqA.fill(0);
  _freqB.fill(0);
  if (buildCharFreqArray(_freqA, a) && buildCharFreqArray(_freqB, b)) {
    let ic = 0;
    for (let i = 0; i < CHAR_FREQ_SIZE; i++) {
      ic += Math.min(_freqA[i], _freqB[i]);
    }
    const total = a.length + b.length;
    return total === 0 ? 1 : (2 * ic) / total;
  }

  // Non-ASCII fallback
  const freqAMap = charFrequencyMap(a);
  const freqBMap = charFrequencyMap(b);

  const ic = intersectCount(freqAMap, freqBMap);
  const total = totalCount(freqAMap) + totalCount(freqBMap);

  if (total === 0) return 1;
  return (2 * ic) / total;
}

/**
 * Sørensen-Dice coefficient based on character n-grams.
 *
 * @param a - First string
 * @param b - Second string
 * @param n - N-gram size (default: 2)
 * @returns Bigram Sørensen-Dice coefficient in [0, 1]
 */
export function sorensenNgram(a: string, b: string, n = 2): number {
  // Fast path: integer-encoded bigrams
  const freqAInt = ngramFrequencyMap(a, n);
  const freqBInt = ngramFrequencyMap(b, n);

  if (freqAInt !== null && freqBInt !== null) {
    const ic = intersectCountInt(freqAInt, freqBInt);
    const total = totalCountInt(freqAInt) + totalCountInt(freqBInt);
    return total === 0 ? 1 : (2 * ic) / total;
  }

  // Fallback: string-keyed n-grams
  const freqA = frequencyMap(ngrams(a, n));
  const freqB = frequencyMap(ngrams(b, n));

  const ic = intersectCount(freqA, freqB);
  const total = totalCount(freqA) + totalCount(freqB);

  if (total === 0) return 1;
  return (2 * ic) / total;
}

File: packages/distance-wasm/src/token/sorensen.rs
/// Sørensen-Dice coefficient — character multiset.
///
/// DSC(A, B) = 2 * |A ∩ B| / (|A| + |B|)
/// Time: O(m + n)

pub fn sorensen(a: &str, b: &str) -> f64 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();

    let mut freq_a = [0u32; 128];
    let mut freq_b = [0u32; 128];

    if !crate::utils::build_char_freq(&mut freq_a, a_bytes) || !crate::utils::build_char_freq(&mut freq_b, b_bytes) {
        return sorensen_fallback(a_bytes, b_bytes);
    }

    let ic = crate::utils::intersect_count(&freq_a, &freq_b);
    let total = a_bytes.len() as u32 + b_bytes.len() as u32;
    if total == 0 { 1.0 } else { 2.0 * ic as f64 / total as f64 }
}

fn sorensen_fallback(a: &[u8], b: &[u8]) -> f64 {
    let mut fa = std::collections::HashMap::<u8, u32>::new();
    let mut fb = std::collections::HashMap::<u8, u32>::new();
    for &c in a { *fa.entry(c).or_insert(0) += 1; }
    for &c in b { *fb.entry(c).or_insert(0) += 1; }

    let (smaller, larger) = if fa.len() <= fb.len() { (&fa, &fb) } else { (&fb, &fa) };
    let mut ic = 0u32;
    for (k, &v) in smaller {
        if let Some(&w) = larger.get(k) { ic += v.min(w); }
    }
    let total = a.len() as u32 + b.len() as u32;
    if total == 0 { 1.0 } else { 2.0 * ic as f64 / total as f64 }
}
